com.unity.shadergraph/Editor/Drawing/Controllers/ShaderInputViewController.cs

```diff
@@ -0,0 +1,316 @@
+        internal IList<ShaderInput> shaderInputsToDelete { get; set; } = new List<ShaderInput>();
+    }
+
+    class ShaderInputViewController : SGViewController<ShaderInput, ShaderInputViewModel>
+    {
+        // Exposed for PropertyView
+        internal GraphData graphData => DataStore.State;
+
+        internal ShaderInputViewController(ShaderInput shaderInput, ShaderInputViewModel inViewModel, GraphDataStore graphDataStore)
+            : base(shaderInput, inViewModel, graphDataStore)
+        {
+            InitializeViewModel();
+
+            m_BlackboardPropertyView = new BlackboardPropertyView(ViewModel);
+            m_BlackboardPropertyView.controller = this;
+            m_BlackboardPropertyView.RegisterCallback<AttachToPanelEvent>((evt) => ViewModel.updateSelectionStateAction(evt));
+            m_BlackboardPropertyView.RegisterCallback<DetachFromPanelEvent>((evt) => ViewModel.persistViewDataKeyAction(evt));
+
+            m_BlackboardRowView = new SGBlackboardRow(m_BlackboardPropertyView, null);
+            m_BlackboardRowView.expanded = SessionState.GetBool($"Unity.ShaderGraph.Input.{shaderInput.objectId}.isExpanded", false);
+        }
+
+        void InitializeViewModel()
+        {
+            ViewModel.Model = Model;
+            ViewModel.IsSubGraph = DataStore.State.isSubGraph;
+            ViewModel.IsInputExposed = (DataStore.State.isSubGraph || (Model.isExposable && Model.generatePropertyBlock));
+            ViewModel.InputName = Model.displayName;
+            switch (Model)
+            {
+                case AbstractShaderProperty shaderProperty:
+                    ViewModel.InputTypeName = shaderProperty.GetPropertyTypeString();
+                    // HACK: Handles upgrade fix for deprecated old Color property
+                    shaderProperty.onBeforeVersionChange += (_) => graphData.owner.RegisterCompleteObjectUndo($"Change {shaderProperty.displayName} Version");
+                    break;
+                case ShaderKeyword shaderKeyword:
+                    ViewModel.InputTypeName = shaderKeyword.keywordType  + " Keyword";
+                    ViewModel.InputTypeName = shaderKeyword.isBuiltIn ? "Built-in " + ViewModel.InputTypeName : ViewModel.InputTypeName;
+                    break;
+            }
+
+            ViewModel.requestModelChangeAction = this.RequestModelChange;
+        }
+
+        SGBlackboardRow m_BlackboardRowView;
+        BlackboardPropertyView m_BlackboardPropertyView;
+
+        internal SGBlackboardRow BlackboardItemView => m_BlackboardRowView;
+
+        protected override void RequestModelChange(IGraphDataAction changeAction)
+        {
+            DataStore.Dispatch(changeAction);
+        }
+
+        // Called by GraphDataStore.Subscribe after the model has been changed
+        protected override void ModelChanged(GraphData graphData, IGraphDataAction changeAction)
+        {
+            if (changeAction is ChangeExposedFlagAction changeExposedFlagAction)
```

Comment:
Will change this (and the similar function in `BlackboardController` and `BlackboardSectionController`) to use a switch statement

Regarding the `DirtyNodes()` thing, one feature (or) limitation (your mileage might vary) of how the controllers currently function is that every controller receives the notifications from the data store for every action that was dispatched. 

The reason why those `DirtyNodes()` calls are specified in all those cases is because there are many more types of `IGraphDataAction`s that do not require the `DirtyNodes()` calls and I am under the impression that dirty-ing all the nodes is a pretty expensive operation, especially with larger graphs, hence I am trying to be more specific about when it is called.

Thinking about how to improve this in the future, it should be pretty straightforward to add some data like an enum to the `GraphDataActions` which specify if its the kind of action that requires dirtying the nodes or not, which in general is something I want to do anyway, i.e. have some manner of scoping for actions so that we can be more specific about our responses and how we handle them.